src/iter/collect/consumer.rs

```diff
@@ -70,44 +112,74 @@ impl<'c, T: Send + 'c> Consumer<T> for CollectConsumer<'c, T> {
 }
 
 impl<'c, T: Send + 'c> Folder<T> for CollectFolder<'c, T> {
-    type Result = ();
+    type Result = PartialCollect<T>;
 
     fn consume(mut self, item: T) -> CollectFolder<'c, T> {
-        // Compute target pointer and write to it. Safe because the iterator
-        // does all the bounds checking; we're only avoiding the target drop.
-        let head = self
-            .target
-            .next()
-            .expect("too many values pushed to consumer");
+        // Compute target pointer and write to it
+
+        // WIP: Does not work yet if T is ZST
+        if self.ptr == self.end {
+            panic!("too many values pushed to consumer");
+        }
         unsafe {
+            let head = self.ptr;
+            self.ptr = self.ptr.add(1);
             ptr::write(head, item);
         }
 
         self.local_writes += 1;
         self
     }
 
-    fn complete(self) {
+    fn complete(mut self) -> Self::Result {
         // NB: We don't explicitly check that the local writes were complete,
         // but `Collect::complete()` will assert the global write count.
 
         // track total values written
         self.global_writes
             .fetch_add(self.local_writes, Ordering::Relaxed);
+        self.complete = true;
+        PartialCollect {
+            start: self.start,
+            len: self.local_writes,
+        }
```

Comment:
There's a problem here if `local_writes` is indeed not covering the whole chunk that the folder was given, then we can't merge this PartialCollect with its right-hand side neighbord.